boto3/s3/transfer.py

```diff
@@ -0,0 +1,448 @@
+        self._amount_read = where
+
+    def close(self):
+        self._fileobj.close()
+
+    def tell(self):
+        return self._amount_read
+
+    def __len__(self):
+        # __len__ is defined because requests will try to determine the length
+        # of the stream to set a content length.  In the normal case
+        # of the file it will just stat the file, but we need to change that
+        # behavior.  By providing a __len__, requests will use that instead
+        # of stat'ing the file.
+        return self._size
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+
+    def __iter__(self):
+        # This is a workaround for http://bugs.python.org/issue17575
+        # Basically httplib will try to iterate over the contents, even
+        # if its a file like object.  This wasn't noticed because we've
+        # already exhausted the stream so iterating over the file immediately
+        # steps, which is what we're simulating here.
+        return iter([])
+
+
+class StreamReaderProgress(object):
+    """Wrapper for a read only stream that adds progress callbacks."""
+    def __init__(self, stream, callback=None):
+        self._stream = stream
+        self._callback = callback
+
+    def read(self, *args, **kwargs):
+        value = self._stream.read(*args, **kwargs)
+        if self._callback is not None:
+            self._callback(len(value))
+        return value
+
+
+class ThreadSafeWriter(object):
+    def __init__(self, write_stream):
+        self._write_stream = write_stream
+        self._lock = threading.Lock()
+
+    def pwrite(self, data, offset):
+        with self._lock:
+            self._write_stream.seek(offset)
+            self._write_stream.write(data)
+
+    def close(self):
+        return self._write_stream.close()
+
+
+class OSUtils(object):
+    def get_file_size(self, filename):
+        return os.path.getsize(filename)
+
+    def open_file_chunk_reader(self, filename, start_byte, size, callback):
+        return ReadFileChunk.from_filename(filename, start_byte,
+                                           size, callback)
+
+    def open(self, filename, mode):
+        return open(filename, mode)
+
+    def wrap_stream_with_callback(self, stream, callback):
+        return StreamReaderProgress(stream, callback)
+
+    def wrap_thread_safe_writer(self, stream):
+        return ThreadSafeWriter(stream)
+
+
+class MultipartUploader(object):
+    def __init__(self, client, config, osutil,
+                 executor_cls=futures.ThreadPoolExecutor):
+        self._client = client
+        self._config = config
+        self._os = osutil
+        self._executor_cls = executor_cls
+
+    def upload_file(self, filename, bucket, key, callback, extra_args):
+        response = self._client.create_multipart_upload(Bucket=bucket,
+                                                        Key=key, **extra_args)
```

Comment:
:+1: At the very least we should document which command the `extra_args` applies to in order to prevent confusion.